PlutoHelperAgent/SharedFunctions.m

```diff
@@ -0,0 +1,236 @@
+                                                             (UInt32)service.length,
+                                                             
+                                                             [service UTF8String],
+                                                             
+                                                             0,
+                                                             
+                                                             NULL,
+                                                             
+                                                             &pwLength,
+                                                             
+                                                             &pwData,
+                                                             
+                                                             &itemRef      // Get a reference this time
+                                                             
+                                                             );
+    
+    if (pwAccessStatus == errSecSuccess) {
+        
+        NSData* data = [NSData dataWithBytes:pwData length:pwLength];
+        
+        NSString* password = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+        
+        SecKeychainAttribute attrs2[] = {
+            
+            { kSecAccountItemAttr, 0, NULL }
+            
+        };
+        
+        SecKeychainAttributeList attributes2 = { 1, attrs2 };
+        
+        OSStatus unAccessStatus = SecKeychainItemCopyContent(itemRef, NULL, &attributes2, NULL, NULL);
+        
+        if (unAccessStatus == errSecSuccess) {
+            
+            NSLog(@"Username retrived from Apple Keychain");
+            
+            NSData* data8 = [NSData dataWithBytes:attributes2.attr->data length:attributes2.attr->length];
+            
+            NSString* usernamedata = [[NSString alloc] initWithData:data8 encoding:NSUTF8StringEncoding];
+            
+            NSArray *returnStrings = [NSArray arrayWithObjects:
+                                      usernamedata,
+                                      password,
+                                      nil];
+            
+            return returnStrings;
+            
+            
+        } else {
+            
+            NSLog(@"Username not retrived from Apple Keychain");
+        }
+        
+    } else {
+        
+        NSLog(@"Keychain read failed: %@", SecCopyErrorMessageString(pwAccessStatus, NULL));
+        
+    }
+    
+    if (pwData) SecKeychainItemFreeContent(NULL, pwData);  // Free memory
+    
+    NSArray *failedReturnStrings = [NSArray arrayWithObjects:
+                                    @"Username not loaded",
+                                    @"Password not loaded",
+                                    nil];
+    
+    return failedReturnStrings;
+    
+}
+
++ (NSString *) login_to_project_server {
```

Comment:
this method seems unduly long, again with response variables like error2 etc.  please could you split it down into individual methods that do 1 thing and call each one from here.

Reply:
Done